src/main/java/com/google/sps/servlets/DeleteFolderServlet.java

```diff
@@ -0,0 +1,109 @@
+
+import com.google.appengine.api.users.UserService;
+import com.google.appengine.api.users.UserServiceFactory;
+import com.google.appengine.api.datastore.DatastoreService;
+import com.google.appengine.api.datastore.DatastoreServiceFactory;
+import com.google.appengine.api.datastore.Entity;
+import com.google.appengine.api.datastore.EntityNotFoundException;
+import com.google.appengine.api.datastore.KeyFactory;
+
+import com.google.appengine.api.datastore.PreparedQuery;
+import com.google.appengine.api.datastore.Query;
+
+import javax.servlet.annotation.WebServlet;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import java.io.IOException;
+import com.google.gson.Gson;
+import com.google.sps.tool.ResponseSerializer;
+import com.google.sps.data.Card;
+import com.google.sps.data.Folder;
+
+@WebServlet("/deletefolder")
+public class DeleteFolderServlet extends HttpServlet {
+
+  @Override
+  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    UserService userService = UserServiceFactory.getUserService();
+    if (!userService.isUserLoggedIn()) {
+      String jsonErrorInfo = ResponseSerializer.getErrorJson("User not logged in");
+      response.setContentType("application/json;");
+      response.getWriter().println(new Gson().toJson(jsonErrorInfo));
+      return;
+    }
+
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+    String folderKey = request.getParameter("folderKey");
+    Entity folder = getExistingFolderInDatastore(datastore, folderKey);
+
+    if (folder == null) {
+      String jsonErrorInfo = ResponseSerializer.getErrorJson("Cannot delete Folder at the moment");
+      response.setContentType("application/json;");
+      response.getWriter().println(new Gson().toJson(jsonErrorInfo));
+    } else {
+      deleteAllCardsInsideFolder(folder);
+      deleteFolderWithRetries(folder);
+    }
+  }
+
+  private Entity getExistingFolderInDatastore(DatastoreService datastore, String cardKey)
+      throws IOException {
+    try {
+      return datastore.get(KeyFactory.stringToKey(cardKey));
+    } catch (EntityNotFoundException e) {
+      return null;
+    }
+  }
+
+  private void deleteFolderWithRetries(Entity folder) {
+    int retries = 5;
+    while (true) {
+      try {
+        Folder.deleteFolder(folder);
+        break;
+      } catch (Exception e) {
+        if (retries == 0) {
+          Folder.addDatastoreDeleteTaskToQueue(folder);
+          break;
+        }
+        --retries;
+      }
+    }
+  }
+
+  private void deleteAllCardsInsideFolder(Entity folder) throws IOException {
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+    Query cardQuery = new Query("Card").setAncestor(folder.getKey());
+    PreparedQuery results = datastore.prepare(cardQuery);
+
+    // TODO(ngothomas): bug with testing local blobstore
```

Comment:
This comment could be improved: specify what bug you're running into, not just "there is a bug". This could be as simple as "blob deletion doesn't work on local blobstore" or "deleting blobs on local blobstore throws exceptions"

Reply:
done